docs/pages/visual-test/Masonry.tsx

```diff
@@ -1,45 +1,132 @@
+    },
+    {
+      color: '#6d6368',
+      naturalHeight: 710,
+      src: 'https://i.ibb.co/zmFd0Dv/stock13.jpg',
+      width: 474,
+      name: 'Border of China and Vietnam',
+    },
+  ];
+
+  const pinList = [...new Array(3)].map(() => [...pins]).flat();
+  return Promise.resolve(pinList);
+}
+
+// Component to display a skeleton pin
+function SkeletonPin({ height }) {
+  const refCallback = useCallback((ref) => {
+    // rewrite parent to use abs vs transform
+    if (ref) {
+      const parentEl = ref.parentElement;
+      if (parentEl.style.transform) {
+        const computedStyles = getComputedStyle(parentEl);
+        if (computedStyles.transform) {
+          const matrix = new DOMMatrix(computedStyles.transform);
+          const { m41: x, m42: y } = matrix;
+          parentEl.style.transform = '';
+          parentEl.style.top = `${y}px`;
+          parentEl.style.left = `${x}px`;
+        }
+      }
+    }
+  }, []);
+  return (
+    <div ref={refCallback} className="SkeletonPin__Loading">
+      <Box height={height} width="100%" />
+    </div>
+  );
+}
+
+function GridComponent({ data }) {
   return (
-    <Box color="successBase" height={data.height} width={50}>
-      <Text>{data.name}</Text>
-    </Box>
+    <Flex direction="column">
+      <Image
+        alt={data.name}
+        color={data.color}
+        naturalHeight={data.naturalHeight}
+        naturalWidth={data.width}
+        src={data.src}
+      />
+    </Flex>
   );
 }
 
+const randomPinHeight = () => Math.random() * 200 + 100;
+
+const skeletonPins = [...new Array(3)]
+  .map(() => [
+    { id: 1, height: randomPinHeight(), width: 474 },
+    { id: 2, height: randomPinHeight(), width: 474 },
+    { id: 3, height: randomPinHeight(), width: 474 },
+    { id: 4, height: randomPinHeight(), width: 474 },
+    { id: 5, height: randomPinHeight(), width: 474 },
+  ])
+  .flat();
+
 export default function Snapshot() {
+  const [pins, setPins] = useState([]);
+  const scrollContainerRef = useRef(null);
+  const gridRef = useRef(null);
+
+  useEffect(() => {
+    setTimeout(() => {
+      getPins().then(setPins);
+    }, 2500);
+  }, []);
+
   return (
     <ColorSchemeProvider colorScheme="light">
-      <Box color="default" display="inlineBlock" padding={1} width={300}>
+      <div
+        ref={scrollContainerRef}
+        style={{ overflowY: 'scroll', height: '100vh', width: '100vw' }}
+      >
         <Masonry
-          columnWidth={50}
-          gutterWidth={3}
-          items={dataObject}
-          minCols={1}
+          ref={(ref) => {
+            gridRef.current = ref;
+          }}
+          columnWidth={170}
+          gutterWidth={20}
+          items={pins}
+          layout="basicCentered"
+          loadingStateItems={skeletonPins}
           renderItem={({ data }) => <GridComponent data={data} />}
+          renderLoadingItems={({ data }) => <SkeletonPin height={data.height} />}
```

Comment:
how configurable do we want this to be? @liuyenwei brought up a good point that this might make it easier to experiment different loading states

Reply:
discussed in https://docs.google.com/document/d/10wCnGJ6uPdtDAAkzLYFJCt-i8rmERE_8PCrFnGu9crQ/edit#heading=h.oc974dez4na